### ISSUE TITLE
pdist with large inputs is giving illegal memory exception on CUDA

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#15511]
## 🐛 Bug

pdist with large inputs is giving illegal memory exception on CUDA

Reported on the forums here: https://discuss.pytorch.org/t/does-pytorch-pdist-have-size-limit/32898

## To Reproduce

```python
import math
import torch #version 1.0
import torch.nn.functional as F

sizes = [100, 1000, 3000, 5000, 7500, 8500, 9999, 10000]
for size in sizes:
    print("[{}] Start".format(size))
    a = torch.randn(size, 20).float().cuda()
    print("Before")
    # The operation
    b = F.pdist(a, p=2)
    print("After")
    a = torch.randn(size, 20).float().cuda()
    print("---End---")
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by thipokKub** [comment]
Here is the output. I've install pytorch using `conda` as stated from the official website.
```
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 396.37
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] Could not collect
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.3                      1  
[conda] mkl_fft                   1.0.6                    py36_0    conda-forge
[conda] mkl_random                1.0.2                    py36_0    conda-forge
[conda] pytorch                   1.0.0           py3.6_cuda9.0.176_cudnn7.4.1_1    pytorch
[conda] torch                     0.4.1                     <pip>
[conda] torchvision               0.2.1                     <pip>
[conda] torchvision               0.2.1                      py_2    pytorch
```
Also this is the cuda and cudnn version `9.0.176` and `7401`

**Comment by ngimel** [comment]
Does not repro on current master, reproes on an older build. Looks like #12302 (merged 8 days ago) fixed whatever bug there was.

**Comment by ngimel** [comment]
Can we reopen this (see edit) and revert #12302? This https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/DistanceKernel.cu#L167 from #12302 would result in silent launch error for all sizes larger than ~350. Bug for 10000 is likely still there. @erikbrinkman, to fix #12302 it would be more robust to have batch size as a gridDim.y, and leave gridDim.x as number of output elements, because max value of gridDim.x is ~2 million, and max value of gridDim.y is 65535. Ideally, you'd want to support batches bigger than 65535 with kernel launches in a loop.

**Comment by ngimel** [comment]
The problem is with this line https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/DistanceKernel.cu#L87-L88, (-1 stops accounting for fp truncation for large enough values, because -1 becomes a no-op). This line is supposed to produce 0 for the first (n-1) k values, then 1 for the next (n-2) values then 2 for the next (n-3) values etc, yet for larger n it stops working. Fast fix (did not check) would be to declare n2 as double, and call device_sqrt<double>, hopefully that will alleviate accuracy issues.

**Comment by erikbrinkman** [comment]
There's a PR here that solves this #12003 for arbitrarily large `n`, but I'm not sure why I didn't think to just mandate the math is always done with doubles. I don't know at what point 64bit floating point will stop being precise enough, but it's at least enough for `2**15`, which already seems excessively large (just verifying the final arrays were identical took 5s on my laptop). The simplest PR would be to switch this to always using a double and probably emit a warning if n is exceedingly large (`2 ** 15`).

The following snippet shows that doubles are precise enough:
```
import torch
n = torch.tensor(2 ** 15, dtype=torch.int64)
truth = torch.cat([torch.full((n - i - 1,), i, dtype=torch.int64) for i in range(n)])
n2 = n.to(torch.float64) - .5
k = torch.arange(n * (n - 1) // 2)
test = (n2 - torch.sqrt(n2 * n2 - 2 * k.to(torch.float64) - 1)).to(torch.int64)
assert torch.all(truth == test)
```

I'd just submit a PR for this since it's simple, but I don't have an active pytorch dev environment anymore, so it might be faster for someone else.

**Comment by ngimel** [comment]
There are currently several issues
1) launch bound errors are not checked
2) setting gridDim.y to number of output elements in forward limits max size to ~360
3) setting gridDim.z to `grid_z = (dist.numel() + block_z - 1) / block_z;` limits max size to ~360*2 ? (I think)
4) Once those are fixed (I don't know how hard it is to fix for backward), next thing is to fix index computation

**Comment by erikbrinkman** [comment]
@ngimel 

1. Can you point to somewhere else in pytorch this is done? At the time I wrote this, kernel launch errors didn't seem to be checked anywhere, and I'd at least like to see it done consistently.
2. I assume you're talking about `n` when you refer to max size (i.e. the number of vectors)? This point seems accurate, but I also feel like I've tested this with 8192 vectors and not had a problem. If it is an issue, it shouldn't be too hard to fix (adding one more loop to the actual kernel).
3. Roughly identical to 2.
4. Agreed.

**Comment by ngimel** [comment]
E.g. here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L264, add cudaGetLastError call after kernel launch. 
If launch parameters are invalid (too many blocks, too many threads etc) it will error out immediately.
2. Yes, I was talking about number of vectors. Rather than adding a loop, it may be easier to set gridDim.x to the number of output elements, because max value of gridDim.x is 2^31.

**Comment by gchanan** [comment]
@ngimel: when linking to the code on github do you mind linking to the fixed hash -- I'm occasionally worried I'm reading the wrong lines because another commit updated them.  At least in chrome you can get this by pressing 'y', which also seems to inline the code. https://github.com/pytorch/pytorch/blob/b1529eeadb531a2454ec8c851f70034b59664df7/aten/src/ATen/native/cuda/DistanceKernel.cu#L87-L88.

**Comment by gchanan** [comment]
closed by https://github.com/pytorch/pytorch/pull/15901.
